services/plugins/npm-plugins-service.ts

```diff
@@ -0,0 +1,49 @@
+import {NpmPluginsSource} from "./npm-plugins-source";
+import {NpmRegistryPluginsSource} from "./npm-registry-plugins-source";
+import {NpmjsPluginsSource} from "./npmjs-plugins-source";
+
+export class NpmPluginsService implements INpmPluginsService {
+	constructor(private $injector: IInjector) { }
+
+	public search(projectDir: string, keywords: string[], modifySearchQuery: (keywords: string[]) => string[]): IFuture<IPluginsSource> {
+		return ((): IPluginsSource => {
+			let npmjsPluginsSource = this.preparePluginsSource(NpmjsPluginsSource, projectDir, keywords).wait();
+
+			if (npmjsPluginsSource.hasPlugins()) {
+				return npmjsPluginsSource;
+			}
+
+			let npmRegistryPluginsSource = this.preparePluginsSource(NpmRegistryPluginsSource, projectDir, keywords).wait();
```

Comment:
you can extract the lines from 16 to 20 to a private method and call them here and in `optimizedSearch` method as well